File: python/day5/__main__.py
def read(filename):
    part = 1
    ranges = []
    available = []
    with open(filename) as f:
        for line in f:
            if line == "\n":
                part = 2
                continue

            if part == 1:
                ranges.append([ int(num) for num in line.split("-") ])

            if part == 2:
                available.append(int(line))
    return ranges, available

def fresh(ranges, item):
    for r in ranges:
        if r[0] <= item and item <= r[1]:
            return True
    return False

def puzzle1(ranges, available):
    count = 0
    for item in available:
        if fresh(ranges, item):
            count+=1
    return count

def puzzle2(ranges):
    merged = [False] * len(ranges)
    def merge(r1, r2):
        # print("TEST", r1, r2)
        if r1[1] < r2[0] or r1[0] > r2[1]:
            return r1, False
        # print("MERGE", r1, r2)
        return [ min(r1[0], r2[0]), max(r1[1], r2[1]) ], True
    for i in range(1, len(ranges)):
        for j in range(i):
            if merged[j]:
                continue
            ranges[i], merged[j] = merge(ranges[i], ranges[j])
    # print(ranges)
    # print(merged)

    sum = 0
    for i in range(len(ranges)):
        r = ranges[i]
        if not merged[i]:
            sum += r[1] - r[0] + 1
    return sum

ranges, available = read("input/day5/puzzle.txt")
# print(ranges, available)
print(puzzle1(ranges, available))
print(puzzle2(ranges))

File: go/day5/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	ranges, available := read("input/day5/puzzle.txt")
	fmt.Println("Puzzle1:", puzzle1(ranges, available))
	fmt.Println("Puzzle2:", puzzle2(ranges))
}

func puzzle1(ranges [][2]int, available []int) int {
	valid := func(ranges [][2]int, item int) bool {
		for _, r := range ranges {
			if r[0] <= item && item <= r[1] {
				return true
			}
		}
		return false
	}
	count := 0
	for _, item := range available {
		if valid(ranges, item) {
			count++
		}
	}
	return count
}

func puzzle2(ranges [][2]int) int {
	merge := func(r1 *[2]int, r2 [2]int) bool {
		if r1[1] < r2[0] || r1[0] > r2[1] {
			return false
		}
		(*r1)[0] = min((*r1)[0], r2[0])
		(*r1)[1] = max((*r1)[1], r2[1])
		return true
	}

	merged := make([]bool, len(ranges))
	for i := 1; i < len(ranges); i++ {
		for j := 0; j < i; j++ {
			if merged[j] {
				continue
			}
			merged[j] = merge(&ranges[i], ranges[j])
		}
	}

	sum := 0
	for i, r := range ranges {
		if !merged[i] {
			sum += r[1] - r[0] + 1
		}
	}
	return sum
}

func read(filename string) ([][2]int, []int) {
	contents, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}

	sections := strings.Split(string(contents), "\n\n")
	section1, section2 := sections[0], sections[1]

	ranges := [][2]int{}
	for _, line := range strings.Split(section1, "\n") {
		dash := strings.Index(line, "-")
		i1, _ := strconv.Atoi(line[:dash])
		i2, _ := strconv.Atoi(line[dash+1:])
		ranges = append(ranges, [2]int{i1, i2})
	}

	available := []int{}
	for _, line := range strings.Split(section2, "\n") {
		if line == "" {
			continue
		}
		i, _ := strconv.Atoi(line)
		available = append(available, i)
	}

	return ranges, available
}
